Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -106,6 +119,55 @@ protected override BoolResult InitializeCore(OperationContext context)
             return result;
         }
 
+        /// <inheritdoc />
+        public override void SetDatabaseMode(bool isDatabaseWriteable)
+        {
+            if (IsGarbageCollectionEnabled != isDatabaseWriteable)

```

Comment:
I don't really like the code that deals with _compactionTimer.

For instance, why in this case there is no checks for IsGarbageCollectionEnabled?
Should the logic be in initializeCore or only in SetDatabaseMode?

Reply:
Because we use _isDatabaseWriteable directly. IsGarbageCollectionEnabled is exactly the same as "Is the database in writeable mode". This logic needs to be here, because this is called whenever a machine is elected as master

---
In reply to: [313644408](https://github.com/microsoft/BuildXL/pull/731#discussion_r313644408) [](ancestors = 313644408)

Reply:
Here is the case:

IsGarbagleCollectionEnabled is false.
We call SetDatabaseMode(true) for master.
In this case we'll enable the timer even though the GC is disabled.

---
In reply to: [313645345](https://github.com/microsoft/BuildXL/pull/731#discussion_r313645345) [](ancestors = 313645345,313644408)